rocksdb_admin/admin_handler.cpp

```diff
@@ -559,30 +567,26 @@ void AdminHandler::async_tm_addS3SstFilesToDB(
   // we couldn't afford to create a new client for every SST file downloading
   // request, which is not even on any critical code path. Otherwise, we will
   // see latency spike when uploading data to production clusters.
-  static const uint32_t s3_download_limit_mb = request->s3_download_limit_mb;
-  static const std::string s3_bucket = request->s3_bucket;
-  static auto s3Util = common::S3Util::BuildS3Util(
-    s3_download_limit_mb, s3_bucket);
+  std::shared_ptr<common::S3Util> local_s3_util;
+  {
+    std::lock_guard<std::mutex> guard(s3_util_lock_);
+    if (s3_util_ == nullptr || should_new_s3_client(*s3_util_, request.get())) {
+      // Request with different ratelimit or bucket has to wait for old
+      // requests to drain.
+      while (s3_util_ != nullptr && s3_util_.use_count() > 1) {
+        LOG(INFO) << "There are other downloads happening, wait "
+                  << kS3UtilRecheckSec << " seconds";
+        std::this_thread::sleep_for(std::chrono::seconds(kS3UtilRecheckSec));
+      }
```

Comment:
If we do not wait, we will potentially end up calling Aws::initApi() more than Aws::ShutdownApi() when there are multiple concurrent requests. Although from the code and asked aws people if that is allowed to do and they said fine for most cases, i think making the calls always occur in pair will be safer.